angr/state_plugins/preconstrainer.py

```diff
@@ -0,0 +1,176 @@
+            for write in self._input_content.writes:
+                for b in write:
+                    self._preconstrain(b, stdin.read_from(1))
+        else:
+            l.error("Preconstrainer currently only supports a string or a TracerPoV as input content.")
+            return
+
+        stdin.seek(0)
+
+        if repair_entry_state_opts:
+            self.state.options |= {o.TRACK_ACTION_HISTORY}
+
+        # add the preconstraints to the actual constraints on the state if we aren't replacing
+        if o.REPLACEMENT_SOLVER not in self.state.options:
+            self.add_constraints(*self.preconstraints)
+
+    def preconstrain_flag_page(self):
+        """
+        Preconstrain the data in the flag page.
+        """
+
+        if not self._preconstrain_flag:
+            return
+
+        if self._magic_content is None:
+            e_msg = "Trying to preconstrain flag page without CGC magic content. "
+            e_msg += "You should have set record_magic flag for Runner dynamic tracing. "
+            e_msg += "For now, nothing will happen."
+            l.warning(e_msg)
+            return
+
+        for b in range(0x1000):
+            self._preconstrain(self._magic_content[b], self.state.cgc.flag_bytes[b])
+
+    def remove_preconstraints(self, to_composite_solver=True, simplify=True):
+        if not (self._preconstrain_input or self._preconstrain_flag):
+            return
+
+        # cache key set creation
+        precon_cache_keys = set()
+
+        for con in self.preconstraints:
+            precon_cache_keys.add(con.cache_key)
+
+        # if we used the replacement solver we didn't add constraints we need to remove so keep all constraints
+        if o.REPLACEMENT_SOLVER in self.state.options:
+            new_constraints = self.state.se.constraints
+        else:
+            new_constraints = filter(lambda x: x.cache_key not in precon_cache_keys, self.state.se.constraints)
+
+
+        if self.state.has_plugin("zen_plugin"):
+            new_constraints = self.state.get_plugin("zen_plugin").filter_constraints(new_constraints)
+
+        if to_composite_solver:
+            self.state.options.discard(o.REPLACEMENT_SOLVER)
+            self.state.options.add(o.COMPOSITE_SOLVER)
+
+        self.state.release_plugin('solver_engine')
+        self.state.add_constraints(*new_constraints)
+
+        l.debug("downsizing unpreconstrained state")
+        self.state.downsize()
+
+        if simplify:
+            l.debug("simplifying solver")
+            self.state.se.simplify()
+            l.debug("simplification done")
+
+        self.state.se._solver.result = None
+
+    def reconstrain(self):
+        """
+        Re-apply preconstraints to improve solver time, hopefully these
+        constraints still allow us to do meaningful things to state.
+        """
+
+        # test all solver splits
+        subsolvers = self.state.se._solver.split()
```

Comment:
It'd be nicer to just have this be `self.state.solver.add(all_the_preconstraints)` -- that way, it would work with the composite solver. The composite solver will sort out which subsolvers to add which constraints to on its own.

Reply:
I don't know why but doing this makes angr eat up all my RAM...